Validation fix descriptions to describe fix to user

huwb opened the pull request:
I realised when clicking the fix button on the depth cache layers validation that its useful to know what the fix will do.

This PR adds a 'mode' to the fix function where it describes the fix action, so the user knows what will change.

Also adds a fix action for material keyword not set, and for dynamic waves disabled errors.

daleeidd commented:
Great idea. I think using another field on the HelpBoxMessage struct and another parameter on the ShowMessage delegate definition to accommodate the action message will be nicer code wise. Thoughts?

daleeidd commented:
Hmmm I think I understand why it was done though. To avoid duplication of action message copy?

huwb commented:
Nope i just didnt think of doing that :). Thats better thanks!

I might have gone a little overboard.. I added a fix description as a param to showMessage. I decided to make it a mandatory arg - whenever we add validation we must directly specify how to fix it. This is used in both the help boxes, the debug logs, and for the fix tooltip (when a fix method is provided). I think this is good, but required fixing up about 50 places in the code.

While doing this i took the opportunity to freshen some of the bits up and added a fix method or two because i could not help myself. Apologies for so much change!

daleeidd reviewed (changes requested):
> I might have gone a little overboard.. I added a fix description as a param to showMessage. I decided to make it a mandatory arg - whenever we add validation we must directly specify how to fix it. This is used in both the help boxes, the debug logs, and for the fix tooltip (when a fix method is provided). I think this is good, but required fixing up about 50 places in the code.

Worth it.
 
> While doing this i took the opportunity to freshen some of the bits up and added a fix method or two because i could not help myself. Apologies for so much change!

That was brutal. But was really worth the effort. Hopefully I got everything.

huwb commented on crest/Assets/Crest/Crest/Scripts/LodData/OceanDepthCache.cs:
```diff
@@ -437,8 +437,8 @@ public bool Validate(OceanRenderer ocean, ValidatedHelper.ShowMessage showMessag
-                        "Depth cache is outdated. Click <i>Populate Cache</i> or re-bake the cache to bring the cache " +
-                        "up-to-date with component changes.",
+                        "Depth cache is outdated.",
+                        "Click Populate Cache or re-bake the cache to bring the cache up-to-date with component changes.",
```
sadly, the tooltip does not support it and displays its as `<i></i>` :/.

huwb replied:
i did consider perhaps using single quotes instead like 'this'

daleeidd replied:
Is stripping the tags an option? We would only need to strip it before adding it to the tooltip since the Undo.RecordObject takes the tooltip too.

daleeidd replied:
I think a regex (untested) like so will handle all tags: `@"<[^<>]*>"`?

huwb replied:
yeah that pretty much works. its not too hard to make it fail, but i think its a rabbit hole to make it really solid, and failures are noncritical i would say.

`condition is <b>cat < bat && dog > frog </b>.`

daleeidd replied:
True. It's pretty naive, but I didn't expect us to use arrows in such a way. We could change to `<[^<>\s]*>` to handle spaces. I think that should cover quite a lot of cases (probably all of ours).

daleeidd replied:
Here is documentation for rich text for reference: https://docs.unity3d.com/Packages/com.unity.ugui@1.0/manual/StyledText.html

If we really needed to we could cover everything since the amount of tags are pretty small. But can leave it until we need it.

daleeidd replied:
I'll resolve this since it covers the issue.

huwb commented on crest/Assets/Crest/Crest/Scripts/OceanValidation.cs:
```diff
@@ -44,45 +47,61 @@ public struct HelpBoxMessage
-        public static void DebugLog(string message, MessageType type, Object @object = null, System.Action<SerializedObject> action = null)
+        public static void DebugLog(string message, string fixDescription, MessageType type, Object @object = null, ValidationFixFunc action = null)
         {
             message = $"Validation: {message} Click this message to highlight the problem object.";
```
ah well caught done

huwb commented on crest/Assets/Crest/Crest/Scripts/Shapes/ShapeGerstnerBatched.cs:
```diff
@@ -880,7 +882,7 @@ public bool Validate(OceanRenderer ocean, ValidatedHelper.ShowMessage showMessag
                 (
                     "There is no spectrum assigned meaning this Gerstner component won't generate any waves.",
-                    ValidatedHelper.MessageType.Warning, this
+                    "Assign a valid spectrum asset to the Spectrum field.", ValidatedHelper.MessageType.Warning, this
```
yeah good shout, done. also moved the fix action argument to its own line everywhere

huwb commented on crest/Assets/Crest/Crest/Scripts/OceanValidation.cs:
```diff
@@ -198,19 +217,30 @@ public void ShowValidationMessages()
 
                                 if (GUILayout.Button(s_fixButtonContent, GUILayout.ExpandWidth(false), GUILayout.ExpandHeight(true)))
                                 {
+                                    // Fix for real
```
thx done

huwb commented on crest/Assets/Crest/Crest/Scripts/OceanValidation.cs:
```diff
@@ -198,19 +217,30 @@ public void ShowValidationMessages()
                             // Fix the issue button.
                             if (message._action != null)
                             {
+                                // Call fix function with null argument to retrieve the resolution info
```
done

huwb commented on crest/Assets/Crest/Crest/Scripts/OceanRenderer.cs:
```diff
@@ -1288,47 +1286,50 @@ public bool Validate(OceanRenderer ocean, ValidatedHelper.ShowMessage showMessag
                 if (ocean.CreateFoamSim)
                 {
-                    showMessage(LodDataMgrFoam.ERROR_MATERIAL_KEYWORD_MISSING, ValidatedHelper.MessageType.Error, ocean.OceanMaterial);
+                    showMessage("Foam not enabled on ocean material and will not be visible.", LodDataMgrFoam.ERROR_MATERIAL_KEYWORD_MISSING_FIX, ValidatedHelper.MessageType.Error, ocean.OceanMaterial);
```
done

huwb commented on crest/Assets/Crest/Crest/Scripts/LodData/RegisterFoamInput.cs:
```diff
@@ -23,16 +23,14 @@ public class RegisterFoamInput : RegisterLodDataInputDisplacementCorrection<LodD
         protected override string ShaderPrefix => "Crest/Inputs/Foam";
 
 #if UNITY_EDITOR
+        protected override string FeatureToggleLabel => "Create Foam Sim";
```
another great catch. done

huwb commented:
Ok i brought back all the text formatting

daleeidd reviewed (approved):
Great improvement. It builds and I didn't see any issues functionally. Before merging, might as well add these changes to the version history? Otherwise, I can do it in master.

huwb commented:
Thanks! I took a stab at release notes but feel free to add/edit.
